### ISSUE TITLE
Can't set custom logger to subclass of ActiveSupport::LogSubscriber

### ISSUE THREAD

**Issue by scnfilho** [rails/rails#42745]
### Steps to reproduce
1. Set a logger using `.logger=` to a custom subclass of `ActiveSupport::LogSubscriber`
2. Try to use that logger or compare its instance

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "activesupport", "6.1.0"
end

require "active_support"
require "active_support/core_ext/object/blank"
require "minitest/autorun"

class MyLogSubscriber < ActiveSupport::LogSubscriber
end

class BugTest < Minitest::Test
  def test_compare_instance
    logger = Logger.new(Tempfile.new)
    MyLogSubscriber.logger = logger

    assert MyLogSubscriber.new.logger == logger
    refute MyLogSubscriber.new.logger == ActiveSupport::LogSubscriber.logger
  end
end
```

### Expected behavior
The [documentation](https://api.rubyonrails.org/classes/ActiveSupport/LogSubscriber.html) doesn't explicitly say this is an option, but it's easy to assume it would be, given it _must_ be set when not in Rails:

> A logger also needs to be set with `ActiveRecord::LogSubscriber.logger=`. This is assigned automatically in a Rails environment.

### Actual behavior
The actual logger is always `Rails.logger` regardless of what is set, unless one sets directly to the class in ActiveSupport using: `ActiveSupport::LogSubscriber.logger=`

If this is the intended behavior, then maybe the documentation should be updated to recommend that setting instead of setting it to the subclass

### System configuration
**Rails version**: `6.1.3.1` 

**Ruby version**: `2.7.2p137`

**Comment by ghiculescu** [comment]
Hmmm, if https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/activesupport/lib/active_support/log_subscriber.rb#L107 changed to `self.class.logger` would it enable this?

**Comment by scnfilho** [comment]
> Hmmm, if
> 
> https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/activesupport/lib/active_support/log_subscriber.rb#L107
> changed to `self.class.logger` would it enable this?

Yes, and I tested that. I was going to point exactly to that line but wanted to be sure it was really an issue first

**Comment by ghiculescu** [comment]
My guess would be nobody's considered subclassing it before. if you think you can make that work, a PR would be great.

**Comment by mbayucot** [comment]
Hi @scnfilho. Are you working on this? If not, can I pick up this issue?

**Comment by scnfilho** [comment]
> Hi @scnfilho. Are you working on this? If not, can I pick up this issue?

Yes, I'm working on it. As it turns out it's not a simple as changing that line to `self.class.logger` - that works for the contrived example I wrote in the issue, but fails existing specs, and even those existing specs are limited because they don't cover the scenario of use within Rails. I'll explain it in more detail in the PR.
